scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java

```diff
@@ -46,18 +50,12 @@
  * @param <F> future type.
  * @param <T> client lookup value type wrapped in a Try.
  */
-public abstract class BaseAsyncLookupDoFn<A, B, C, F, T> extends DoFn<A, KV<A, T>>
+public abstract class BaseAsyncLookupDoFn<A, B, C, F, T>
+    extends DoFnWithResource<A, KV<A, T>, Pair<C, Cache<A, B>>>
     implements FutureHandlers.Base<F, B> {
   private static final Logger LOG = LoggerFactory.getLogger(BaseAsyncLookupDoFn.class);
 
-  // DoFn is deserialized once per CPU core. We assign a unique UUID to each DoFn instance upon
-  // creation, so that all cloned instances share the same ID. This ensures all cores share the
-  // same Client and Cache.
-  private static final ConcurrentMap<UUID, Object> client = new ConcurrentHashMap<>();
-  private static final ConcurrentMap<UUID, Cache> cache = new ConcurrentHashMap<>();
-  private final UUID instanceId;
```

Comment:
client and cache are basically resources. Delegate logic to `DoFnWithResource` instead